### ISSUE TITLE
potential pitfall: PDA constraint check doesn't validate bump against `findProgramAddress`

### ISSUE THREAD

**Issue by kklas** [project-serum/anchor#978]
Suppose we have the following accounts struct for some kind of initialization instruction in a program:

```
#[derive(Accounts)]
#[instruction(authority_bump: u8)]
pub struct Initialize<'info> {
    #[account(init, payer=payer, space = <space>)]
    state: Account<'info, State>,

    #[account(
        seeds = [
            state.key().as_ref()
        ],
        bump = authority_bump,
    )]
    authority: AccountInfo<'info>,

    payer: Signer<'info>,
    system_program: Program<'info, System>,
}
```

and suppose the `state` address is such that the authority PDA can be derived with say 254 and 250 bump seeds.

The `#[account( seeds = [ state.key().as_ref() ], bump = authority_bump )]` constraint check will successfully validate for authority addresses derived from any of these seeds.

I believe this is dangerous because most users will assume that the check will assert that the bump is the same as the one from calling `findProgramAddress`. This would break frontend client assumptions because now the `authority` is not "stateless" anymore as the clients can't deterministically derive it by using `findProgramAddress` but would first have to fetch the correct bump seed from the state account.

I think the default should be that the `#[account( seeds = [ state.key().as_ref() ], bump = authority_bump )]` constraint check validates the seed against `findProgramAddress` as this is the most common use case and what most developers will expect.
For the cases where other bump seeds can be valid there we could have a flag (e.g. `any_pda_is_valid`):
```
    #[account(
        seeds = [
            state.key().as_ref()
        ],
        bump = authority_bump,
        any_pda_is_valid,
    )]
```

**Comment by armaniferrante** [comment]
The current behavior is that `find_program_address` is used to validate the bump in two cases.

1) Anytime the account is initialized. E.g., `#[account(init, seeds = <>, bump = <>)]`
2) Anytime the bump value is omitted. E.g., `#[account(seeds = <>, bump)]`

While `create_program_address` is used in the third case.

3) `#[account(seeds = <>, bump = <>)]`

The rationale here is that `find_program_address` is expensive, and so we want to avoid using it whenever possible. When 1) is used,  `#[account(seeds = <>, bump = <>)]` is safe because it can only be initialized with a find_program_address check. For 2), it's safe since it uses `find_program_address`. For 3), without `init`, it's not necessarily the case that the bump is the same as the one used by `find_program_address`, and so the best practice is to store the bump in the account on creation, which is what one should do with non Anchor programs.

In short, I believe the current API satisfies your use case. To run find_program_address every time, use option 2).

**Comment by kklas** [comment]
I didn't know you can call it with `#[account(seeds = <>, bump)]`. I've tried with `#[account(seeds = <>)]` but it complained so I assumed the bump needs to be there.

But in any case, 2. doesn't fully satisfy my use case because I want to store the bump in the state during initialization so I can validate the authority in other instructions by loading it from there and doing the cheaper `create` check. This is not convenient because if I use 2., during initialize, I don't know what the bump is from within the instruction (it's not passed as an argument in this case). So I would again have to run `findProgramAddress` to retrieve the bump which is a waste.
The only other way I could do it is just forgo the `#[account()]` macro on the address and do the `find` from the instruction but seems like an anti pattern since the purpose of the whole accounts structs is to not to have to do account validation from instruction.

I'd argue that this use case is too common to not have it be default vs. the `create_program_address` style check which I consider to more of a rare/special use case (there isn't even a default method find PDAs beyond the first bump).

Also, the current implementation is a bit confusing IMO with all the special cases `seeds = <>, bump`, `seeds = <>, bump =<>`, these two with combination with `init`... It's not at all obvious when it will do the `find` check vs the `create` and it creates a bit of an mental overhead when reading the contract trying to remember which is which. 

With the `any_pda_is_valid` approach it would be much more streamlined - if it's present, do the `create` check, if it's not do the `find` check, and it would be consistent accross `init` and no `init` flags. Much more intuitive IMO.

**Comment by paul-schaaf** [comment]
@armaniferrante @kklas 

regarding 1) from armani: are you sure `find_program_address` is always used when initializing? In fact, in the example you used, I think `find_program_address` is not used https://github.com/project-serum/anchor/blob/11764a1fa5ef9daaedd8fcca81e3059d78fd81fa/lang/syn/src/codegen/accounts/constraints.rs#L314

So with the above in mind I think it's actually simpler than armani described. `bump = <>` is not a check but an assigment, so because we are assigning a bump, we know that `create` will be used. If we are not assigning a bump (just `bump`), `find` will be used. So there a just two cases and whether u use init or not has no effect on whether find or create will be used, only the existence of the bump assigment matters.



I'd prefer too if the API was `#[account(seeds = <>)]` instead of `#[account(seeds = <>, bump)]` in the case of no bump assignment. The latter is very unintuitive imo.

I agree it would be nice to have access to the result of the `find` call inside the program without having to call `find` again.

**Comment by armaniferrante** [comment]
@paul-schaaf the bump check is here https://github.com/project-serum/anchor/blob/master/lang/syn/src/codegen/accounts/constraints.rs#L336-L344. It's only run on `init`, even if the bump value is explicitly given.

**Comment by armaniferrante** [comment]
> This would make intuitive sense to me. What are the reasons against this?

There's no reason to use any bump other than the one given by `find_program_address` (at least not that I"m aware of). And so this check is added for safety reasons, as explained in the original post. That is, without you, one could accidentally allow two different accounts to be created with two different bumps, when only one was meant to be created.

**Comment by paul-schaaf** [comment]
one other idea would be to forbid` bump = target` inside the macro when there's an also an `init` in the constraint. if it's always find_program_address anyway, why even give the option to pass in a target if that target has to be the result of find_program_address.

then, we can wait until a user complains that there is no option for non-chanonical seeds at which point we construct a new constraint which takes care of that (e.g. target_non_can)

**Comment by kklas** [comment]
To summarize my suggestion:

**1. Init with *find* check**

a) bump is passed as an instruction argument (and thus can be used inside the instruction method) - **not possible with current API**
```
    #[account(
        init,
        seeds = <>,
        bump = authority_bump,
    )]
```

b) bump is *not* passed as an argument
```
    #[account(
        init,
        seeds = <>
    )]
```

**2. Init with *create* check**

bump must be passed as an instruction argument - **not possible with current API**
```
    #[account(
        init,
        seeds = <>,
        bump = authority_bump,
        any_pda_is_valid
    )]
```

**3. *find* check**

a) bump is passed as an argument - **not possible with current API since it will do a *create* check**
```
    #[account(
        seeds = <>,
        bump = authority_bump,
    )]
```

b) bump is not passed as an argument
```
    #[account(
        seeds = <>
    )]
```

**4. *create* check**

bump must be passed as an argument
```
    #[account(
        seeds = <>,
        bump = authority_bump,
        any_pda_is_valid,
    )]
```


This way we have:
- consistent API
- most common use case as default
- all init/find/create use cases covered

**Comment by akutruff** [comment]
I just independently ran across this issue, and to my surprise and horror, a valid PDA bump after the first expected bump succeeded.  (The case above where `create_program_address` is used.)  If I shipped this way, then all funds that would have gone to the platform could have simply been sent to a blackhole PDA by an attacker.  

After reading this thread, I searched for how this should be done and found the recommendation in: [sea level attacks](https://github.com/project-serum/sealevel-attacks/blob/master/programs/7-bump-seed-canonicalization/recommended/src/lib.rs)

```rust
#[derive(Accounts)]
#[instruction(key: u64)]
pub struct BumpSeed<'info> {
    #[account(seeds = [key.to_le_bytes().as_ref()], bump = data.bump)]
    data: Account<'info, Data>,
}

#[account]
pub struct Data {
    value: u64,
    bump: u8,
}
```

I had no idea that form of verification was even possible, but it does make sense now.

Something like @kklas suggestion does feel more like 'shipping a secure default' than the current API.  Though these new defaults are slower performance-wise, something like `any_pda_is_valid` provides that optimization later on.

**Comment by armaniferrante** [comment]
The current state of this as of v0.21.0 (the next version) will be as follows

- Init with a bump target, i.e. `#[account(init, seeds=[...], bump = <target>)]`, will be a compiler error. Instead one should *always* use bump without a target, i.e. `#[account(init, seeds=[...], bump)]`, which will run find_program_address. If the bump is needed one can use `ctx.bumps.get("<my-account-name>")`. In no cases, should the developer ever pass a bump in to instruction data.
- On all other uses of `#[account(seeds = [...], bump = <target>)]`, my comment above still holds. That is, create_program_address will be used when the `<target>` is provided. And `find_program_address` will be used without a target. In all cases, it's recommended to store the bump on the account data during `init` and then later use that stored bump as  a target for subsequent seeds constraints. If one does this, then all uses will be safe.

**Comment by akutruff** [comment]
Thanks, that sound like an improvement.  So instructions will look like this to store bumps?  It would be nice if it wasn't a string lookup, but I assume there are reasons.

```rust
 pub fn instruction(
        ctx: Context<MyContext>,       
    ) -> ProgramResult {
        let account_a = &mut ctx.accounts.account_a;
        account_a.bump = ctx.bumps.get("account_a");

        let account_b = &mut ctx.accounts.account_b;
        account_b.bump = ctx.bumps.get("account_b");

        Ok(())
    }
```

**Comment by armaniferrante** [comment]
@akutruff correct (though note that `.get()` returns an `Option<u8>`). The string lookup can probably be replaced if we implement PartialEq on all the main account types.
